// @dazl/component-gallery/dist/route/index.js
import { jsxs, jsx } from "react/jsx-runtime";
import { useCallback, memo, useState, lazy, Suspense, Component, useMemo, useEffect, useRef } from "react";
import { z } from "zod";
const componentGallery = "componentGallery";
const styles$1 = {
  componentGallery
};
const useLinkBehaviorOverride = () => {
  return useCallback((container) => {
    if (!container) return;
    const handleClick = (event) => {
      if (!(event.target instanceof Element)) return;
      const link = event.target.closest("a");
      if (!link) return;
      const href = link.getAttribute("href");
      if (!href) return;
      event.preventDefault();
      if (isAbsoluteUrl(href)) {
        window.open(href, "_blank", "noopener,noreferrer");
      }
    };
    container.addEventListener("click", handleClick);
    return () => {
      container.removeEventListener("click", handleClick);
    };
  }, []);
};
const isAbsoluteUrl = (url) => /^[a-z][a-z0-9+.-]*:/i.test(url);
const exampleCard = "exampleCard";
const exampleCardHeader = "exampleCardHeader";
const exampleCardContent = "exampleCardContent";
const exampleCardLoading = "exampleCardLoading";
const exampleCardSpinner = "exampleCardSpinner";
const exampleCardError = "exampleCardError";
const exampleCardErrorIcon = "exampleCardErrorIcon";
const exampleCardErrorMessage = "exampleCardErrorMessage";
const styles = {
  exampleCard,
  exampleCardHeader,
  exampleCardContent,
  exampleCardLoading,
  exampleCardSpinner,
  exampleCardError,
  exampleCardErrorIcon,
  exampleCardErrorMessage
};
const ExampleCard = memo(function Example({ example }) {
  const [Component2] = useState(() => lazy(() => import(
    /* @vite-ignore */
    `/${example.relativePath}`
  )));
  const contentRef = useLinkBehaviorOverride();
  return /* @__PURE__ */ jsxs("div", { id: example.relativePath, className: styles.exampleCard, children: [
    /* @__PURE__ */ jsx("h3", { className: styles.exampleCardHeader, children: example.displayName }),
    /* @__PURE__ */ jsx(ErrorBoundary, { fallback: (error) => /* @__PURE__ */ jsx(ExampleCardError, { error }), children: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(ExampleCardLoading, {}), children: /* @__PURE__ */ jsx("div", { ref: contentRef, className: styles.exampleCardContent, children: /* @__PURE__ */ jsx(Component2, {}) }) }) })
  ] });
});
const ExampleCardLoading = () => {
  return /* @__PURE__ */ jsx("div", { className: styles.exampleCardLoading, children: /* @__PURE__ */ jsx("div", { className: styles.exampleCardSpinner }) });
};
const ExampleCardError = ({ error }) => {
  return /* @__PURE__ */ jsxs("div", { className: styles.exampleCardError, children: [
    /* @__PURE__ */ jsx("span", { className: styles.exampleCardErrorIcon, children: "⚠" }),
    /* @__PURE__ */ jsx("span", { className: styles.exampleCardErrorMessage, children: error.message })
  ] });
};
class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { error: null };
  }
  static getDerivedStateFromError(error) {
    return { error };
  }
  render() {
    if (this.state.error) {
      return this.props.fallback(this.state.error);
    }
    return this.props.children;
  }
}
const scrollExampleCardIntoView = (id) => {
  const element = document.getElementById(id);
  if (!element) return;
  element.scrollIntoView({
    behavior: "instant",
    block: "nearest"
  });
  element.animate(
    [
      { boxShadow: "0 0 0 0 #3E63DD" },
      { boxShadow: "0 0 0 4px #3E63DD", offset: 0.2 },
      { boxShadow: "0 0 0 0 #3E63DD" }
    ],
    {
      duration: 900,
      easing: "ease-out"
    }
  );
};
const componentExampleInfoSchema = z.object({
  displayName: z.string(),
  relativePath: z.string()
});
const componentGalleryParamsSchema = z.object({
  columns: z.number().default(2),
  examples: z.array(componentExampleInfoSchema).default([]),
  selectedExamplePath: z.string().optional(),
  random: z.number().optional()
});
const useLocationHash = () => {
  const [hash, setHash] = useState(() => typeof window !== "undefined" ? window.location.hash : "");
  useEffect(() => {
    const handleHashChange = () => setHash(window.location.hash);
    window.addEventListener("hashchange", handleHashChange);
    return () => window.removeEventListener("hashchange", handleHashChange);
  }, []);
  return hash;
};
const useComponentGalleryParams = () => {
  const hash = useLocationHash();
  return useMemo(() => {
    try {
      const parsed = JSON.parse(decodeURIComponent(hash.slice(1)));
      return componentGalleryParamsSchema.parse(parsed);
    } catch {
      return { columns: 1, examples: [] };
    }
  }, [hash]);
};
const useScrollToExampleCard = (relativePath, counter) => {
  const containerRef = useRef(null);
  useEffect(() => {
    const container = containerRef.current;
    if (!relativePath || !container) return;
    let scrolledManually = false;
    let scrollingProgrammatically = false;
    let animationFrameId = 0;
    const handleScroll = () => {
      if (scrollingProgrammatically) return;
      scrolledManually = true;
    };
    const resizeObserver = new ResizeObserver(() => {
      if (scrolledManually) return;
      scrollingProgrammatically = true;
      scrollExampleCardIntoView(relativePath);
      cancelAnimationFrame(animationFrameId);
      animationFrameId = requestAnimationFrame(() => {
        scrollingProgrammatically = false;
      });
    });
    window.addEventListener("scroll", handleScroll);
    resizeObserver.observe(container);
    return () => {
      window.removeEventListener("scroll", handleScroll);
      resizeObserver.disconnect();
      cancelAnimationFrame(animationFrameId);
    };
  }, [relativePath, counter]);
  return containerRef;
};
function ComponentGallery() {
  const { columns, examples, selectedExamplePath, random } = useComponentGalleryParams();
  const ref = useScrollToExampleCard(selectedExamplePath, random);
  return /* @__PURE__ */ jsx("div", { ref, className: styles$1.componentGallery, style: { "--columns": columns }, children: examples.map((example) => /* @__PURE__ */ jsx(ExampleCard, { example }, example.relativePath)) });
}
export {
  ComponentGallery as default
};
